### PR TITLE
Drop shipping

### PR DESCRIPTION
**Pull request by saurabh6790**
- [x] One receipt of order from Customer, Sales Order will be created, of type Drop Shipment.
- [x] Purchase Order will be created as Shipping Order.
- [x] In Purchase Order, details will be fetched from Sales Order. Sales Order details like Customer details, Item and its selling price will be fetched into Purchase Order.
- [x] Purchase Order of type Shipping Order will carry Customer and Supplier details.

### CODE REVIEW

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.js [hunk 1]
```diff
@@ -50,6 +50,13 @@ erpnext.accounts.SalesInvoiceController = erpnext.selling.SellingController.exte
 		if(doc.update_stock) this.show_stock_ledger();
 
 		if(doc.docstatus==1 && !doc.is_return) {
+			
+			var flag_delivery_note = false;
```
Comment (nabinhait) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.js]: `flag_delivery_note` does not mean anything, use `is_drop_ship` instead

File: erpnext/buying/doctype/purchase_common/purchase_common.py [hunk 2]
```diff
@@ -77,11 +77,13 @@ def validate_for_items(self, obj):
...
-		stopped = frappe.db.sql("""select name from `tab%s` where name = %s and
-			status = 'Stopped'""" % (doctype, '%s'), docname)
-		if stopped:
-			frappe.throw(_("{0} {1} status is 'Stopped'").format(doctype, docname), frappe.InvalidStatusError)
+	def check_for_stopped_or_closed_status(self, doctype, docname):
+		status = frappe.db.get_value(doctype, docname, "status")
+		
+		if status == "Stopped":
+			frappe.throw(_("{0} {1} status is Stopped").format(doctype, docname), frappe.InvalidStatusError)
+		if status == "Closed":
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_common/purchase_common.py]: if status in ("Stopped", "Closed"):

File: erpnext/buying/doctype/purchase_order/purchase_order.js [hunk 3]
```diff
@@ -223,6 +225,18 @@ cur_frm.cscript['Unstop Purchase Order'] = function() {
 	}
 }
 
+cur_frm.cscript['Close Purchase Order'] = function() {
+	var doc = cur_frm.doc;
+	var check = confirm(__("Do you really want to Close ") + doc.name);
+
+	if (check) {
+		return $c('runserverobj', args={'method':'update_status', 'arg': 'Closed', 'docs':doc}, function(r,rt) {
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]: Don't use runserverobj(), use frappe.call()

Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]: You can commonify the calling function for close and stop

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 4]
```diff
@@ -176,8 +184,16 @@ def on_submit(self):
 		purchase_controller.update_last_purchase_rate(self, is_submit = 1)
 
 	def on_cancel(self):
+		if self.is_drop_ship == 1:
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]: write a function and call from on_submit and on_cancel

File: erpnext/controllers/status_updater.py [hunk 5]
```diff
@@ -210,6 +212,16 @@ def _update_children(self, args):
 	def _update_percent_field(self, args):
 		"""Update percent field in parent transaction"""
 		unique_transactions = set([d.get(args['percent_join_field']) for d in self.get_all_children(args['source_dt'])])
+
```
Comment (nabinhait) [inline erpnext/controllers/status_updater.py]: Write in specific controller

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 6]
```diff
@@ -145,6 +165,43 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
...
+			]
+		});
+
+		dialog.fields_dict.proceed.$input.click(function() {
+			args = dialog.get_values();
+			if(!args) return;
+			dialog.hide();
+			return frappe.call({
+				type: "GET",
+				method: "erpnext.selling.doctype.sales_order.sales_order.make_drop_shipment",
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: make_purchase_order_for_drop_shipment

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 7]
```diff
@@ -15,9 +15,20 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
 	refresh: function(doc, dt, dn) {
 		this._super();
 		this.frm.dashboard.reset();
-
+		var flag_drop_ship = false;
+		var flag_delivery_note = false;
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: rename variable

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 8]
```diff
@@ -198,6 +256,22 @@ cur_frm.cscript['Unstop Sales Order'] = function() {
 	}
 }
 
+cur_frm.cscript['Close Sales Order'] = function(){
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: commonify with stop, if possible commonify for sales and purchase

File: erpnext/selling/doctype/sales_order/sales_order.json [hunk 9]
```diff
@@ -2040,6 +2042,29 @@
    "allow_on_submit": 0, 
    "bold": 0, 
    "collapsible": 0, 
+   "fieldname": "per_ordered", 
+   "fieldtype": "Percent", 
+   "hidden": 0, 
+   "ignore_user_permissions": 0, 
+   "in_filter": 1, 
+   "in_list_view": 1, 
+   "label": "%  Ordered",
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.json]: double space

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 10]
```diff
@@ -147,6 +148,11 @@ def update_prevdoc_status(self, flag):
 				doc.set_status(update=True)
 				doc.update_opportunity()
 
+	def validate_drop_ship(self):
+		for d in self.get('items'):
+			if d.is_drop_ship and not d.supplier:
+				frappe.throw(_("#{0} Set Supplier for item {1}").format(d.idx, d.item_code))
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Row #{0}:

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 11]
```diff
@@ -350,7 +359,7 @@ def update_item(source, target, source_parent):
 				"parent": "against_sales_order",
 			},
 			"postprocess": update_item,
-			"condition": lambda doc: doc.delivered_qty < doc.qty
+			"condition": lambda doc: doc.delivered_qty < doc.qty and (doc.is_drop_ship!=1 and not doc.supplier)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Check only for `doc.is_drop_ship!=1`

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 12]
```diff
@@ -378,6 +387,7 @@ def set_missing_values(source, target):
 		target.run_method("calculate_taxes_and_totals")
 
 	def update_item(source, target, source_parent):
+		target.supplier = source.supplier
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Supplier is not needed in invoice

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 13]
```diff
@@ -488,3 +498,93 @@ def get_events(start, end, filters=None):
...
+				["delivery_date", "schedule_date"]
+			],
+			"postprocess": update_item,
+			"condition": lambda doc: doc.ordered_qty < doc.qty and doc.supplier == for_supplier or (doc.is_drop_ship==1 and doc.supplier == for_supplier)
+		},
+		"Sales Taxes and Charges": {
+			"doctype": "Purchase Taxes and Charges",
+			"add_if_empty": True
+		}
+	}, target_doc, postprocess)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: directly call set_missing_values()

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 14]
```diff
@@ -488,3 +498,93 @@ def get_events(start, end, filters=None):
...
+@frappe.whitelist()
+def make_drop_shipment(source_name, for_supplier, target_doc=None):
+	def postprocess(source, target):
+		set_missing_values(source, target)
+		
+	def set_missing_values(source, target):
+		target.supplier = for_supplier
+		target.buying_price_list = frappe.get_value("Supplier", for_supplier, "default_price_list") or 'Standard Buying'
+		target.address_display = ""
+		target.contact_display = ""
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Use `fields_no_map`

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 15]
```diff
@@ -488,3 +498,93 @@ def get_events(start, end, filters=None):
...
+		target.is_drop_ship = 1
+		target.run_method("set_missing_values")
+		target.run_method("calculate_taxes_and_totals")
+
+	def update_item(source, target, source_parent):
+		target.schedule_date = source_parent.delivery_date
+		target.rate = ''
+		target.price_list_rate = ''
+		target.base_amount = (flt(source.qty) - flt(source.ordered_qty)) * flt(source.base_rate)
+		target.amount = (flt(source.qty) - flt(source.ordered_qty)) * flt(source.rate)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Don't set base_amount and amount

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 16]
```diff
@@ -488,3 +498,93 @@ def get_events(start, end, filters=None):
...
+			"doctype": "Purchase Order Item",
+			"field_map":  [
+				["name", "prevdoc_detail_docname"],
+				["parent", "prevdoc_docname"],
+				["parenttype", "prevdoc_doctype"],
+				["uom", "stock_uom"],
+				["delivery_date", "schedule_date"]
+			],
+			"postprocess": update_item,
+			"condition": lambda doc: doc.ordered_qty < doc.qty and doc.supplier == for_supplier or (doc.is_drop_ship==1 and doc.supplier == for_supplier)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: No need of ` or (doc.is_drop_ship==1 and doc.supplier == for_supplier)`

File: erpnext/selling/doctype/sales_order/test_sales_order.py [hunk 17]
```diff
@@ -295,6 +295,45 @@ def test_auto_insert_price(self):
...
+
+		so = make_sales_order(item_list=so_items)
+		po = make_drop_shipment(so.name, '_Test Supplier')
+		dn = make_delivery_note(so.name)
+
+		self.assertEquals(so.customer, po.customer)
+		self.assertEquals(po.items[0].prevdoc_doctype, "Sales Order")
+		self.assertEquals(po.items[0].prevdoc_docname, so.name)
+		self.assertEquals(po.items[0].item_code, po_item.item_code)
+		self.assertEquals(dn.items[0].item_code, dn_item.item_code)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/test_sales_order.py]: - Test cases to check ordered_qty and per_ordered
- Check length of po_items
- Logic of Closed status with reserved and ordered qty

File: erpnext/selling/doctype/sales_order_item/sales_order_item.json [hunk 18]
```diff
@@ -686,6 +686,95 @@
...
+   "hidden": 0, 
+   "ignore_user_permissions": 0, 
+   "in_filter": 0, 
+   "in_list_view": 0, 
+   "label": "Supplier", 
+   "no_copy": 0, 
+   "options": "Supplier", 
+   "permlevel": 0, 
+   "precision": "", 
+   "print_hide": 0,
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order_item/sales_order_item.json]: set as 1

File: erpnext/selling/doctype/sales_order_item/sales_order_item.json [hunk 19]
```diff
@@ -686,6 +686,95 @@
...
+   "fieldtype": "Check", 
+   "hidden": 0, 
+   "ignore_user_permissions": 0, 
+   "in_filter": 0, 
+   "in_list_view": 0, 
+   "label": "Is Drop Ship Item", 
+   "no_copy": 0, 
+   "permlevel": 0, 
+   "precision": "", 
+   "print_hide": 0,
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order_item/sales_order_item.json]: set as 1

File: erpnext/stock/get_item_details.py [hunk 20]
```diff
@@ -31,6 +31,7 @@ def get_item_details(args):
 			"transaction_type": "selling",
 			"ignore_pricing_rule": 0/1
 			"project_name": "",
+			"default_supplier":""
```
Comment (nabinhait) [inline erpnext/stock/get_item_details.py]: remove

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 21]
```diff
@@ -488,3 +498,93 @@ def get_events(start, end, filters=None):
...
 	return data
+
+@frappe.whitelist()
+def make_drop_shipment(source_name, for_supplier, target_doc=None):
+	def postprocess(source, target):
+		set_missing_values(source, target)
+		
+	def set_missing_values(source, target):
+		target.supplier = for_supplier
+		target.buying_price_list = frappe.get_value("Supplier", for_supplier, "default_price_list") or 'Standard Buying'
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: if default_price_list for supplier exists, then only map

**Comment by nabinhait** [comment]
Add screenshots for print format and SO and PO form

**Comment by nabinhait** [comment]
Documentation

**Comment by nabinhait** [comment]
Change Log

File: erpnext/controllers/status_updater.py (hunk 5 above)
Comment (saurabh6790) [inline erpnext/controllers/status_updater.py]: Pending

File: erpnext/selling/doctype/sales_order/test_sales_order.py (hunk 17 above)
Comment (saurabh6790) [inline erpnext/selling/doctype/sales_order/test_sales_order.py]: pending

File: erpnext/accounts/doctype/sales_invoice_item/sales_invoice_item.json [hunk 22]
```diff
@@ -684,6 +684,51 @@
    "allow_on_submit": 0, 
    "bold": 0, 
    "collapsible": 1, 
+   "collapsible_depends_on": "eval:doc.delivered_by_supplier==1", 
+   "fieldname": "by_supplier",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/sales_invoice_item/sales_invoice_item.json]: Change to supplier

File: erpnext/buying/doctype/purchase_order/purchase_order.js [hunk 23]
```diff
@@ -155,6 +161,29 @@ erpnext.buying.PurchaseOrderController = erpnext.buying.BuyingController.extend(
 				frappe.set_route("Form", doclist[0].doctype, doclist[0].name);
 			}
 		});
+	},
+	stop_purchase_order: function(){
+		cur_frm.cscript.update_status('Stop', 'Stopped')
+	},
+	unstop_purchase_order: function(){
+		cur_frm.cscript.update_status('UNSTOP', 'Submitted')
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]: Title case

File: erpnext/buying/doctype/purchase_order/purchase_order.json [hunk 24]
```diff
@@ -151,6 +151,53 @@
    "allow_on_submit": 0, 
    "bold": 0, 
    "collapsible": 0, 
+   "depends_on": "eval:doc.is_drop_ship==1",
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.json]: eval:doc.delivered_by_supplier==1

File: erpnext/buying/doctype/purchase_order/purchase_order.json [hunk 25]
```diff
@@ -151,6 +151,53 @@
...
+   "reqd": 0, 
+   "search_index": 0, 
+   "set_only_once": 0, 
+   "unique": 0
+  }, 
+  {
+   "allow_on_submit": 0, 
+   "bold": 0, 
+   "collapsible": 0, 
+   "depends_on": "eval:doc.is_drop_ship==1",
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.json]: eval:doc.delivered_by_supplier==1

File: erpnext/buying/doctype/purchase_order/purchase_order.json [hunk 26]
```diff
@@ -1404,6 +1561,30 @@
    "allow_on_submit": 0, 
    "bold": 0, 
    "collapsible": 0, 
+   "depends_on": "eval:doc.is_drop_ship==1",
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.json]: eval:doc.delivered_by_supplier==1

File: erpnext/buying/doctype/purchase_order/purchase_order.json [hunk 27]
```diff
@@ -1404,6 +1561,30 @@
    "allow_on_submit": 0, 
    "bold": 0, 
    "collapsible": 0, 
+   "depends_on": "eval:doc.is_drop_ship==1", 
+   "fieldname": "customer_address",
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.json]: Remove customer_address and customer_contact

File: erpnext/buying/doctype/purchase_order/purchase_order.json [hunk 28]
```diff
@@ -329,6 +464,28 @@
   {
    "allow_on_submit": 0, 
    "bold": 0, 
+   "collapsible": 0, 
+   "fieldname": "delivered_by_supplier",
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.json]: Make read only

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 29]
```diff
@@ -325,3 +339,39 @@ def make_stock_entry(purchase_order, item_code):
...
+	po = frappe.get_doc("Purchase Order", name)
+	po.update_status(status)
+	
+@frappe.whitelist()
+def	delivered_by_supplier(purchase_order):
+	po = frappe.get_doc("Purchase Order", purchase_order)
+	update_delivered_qty(po)
+	po.update_status("Delivered")
+	
+def update_delivered_qty(purchase_order):
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]: Move to sales order controller and use select sum()

File: erpnext/buying/doctype/purchase_order/purchase_order.py [hunk 30]
```diff
@@ -325,3 +339,39 @@ def make_stock_entry(purchase_order, item_code):
...
+			frappe.db.sql(""" update `tabSales Order Item` 
+				set delivered_qty = (ifnull(delivered_qty, 0) + %(qty)s)
+				where item_code='%(item_code)s' and parent = '%(name)s' 
+			"""%{"qty": item.qty, "item_code": item.item_code, "name": item.prevdoc_docname})
+		
+			sales_order_list.append(item.prevdoc_docname)
+			
+	update_per_delivery(sales_order_list)
+		
+def update_per_delivery(sales_order_list):
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/purchase_order.py]: moved to so

File: erpnext/buying/doctype/purchase_order/test_purchase_order.py [hunk 31]
```diff
@@ -70,6 +70,20 @@ def test_uom_integer_validation(self):
...
+	
+	def test_ordered_qty_for_closing_po(self):
+		from erpnext.stock.doctype.item.test_item import make_item
+		
+		item = make_item("_Test Close Item", {"is_stock_item": 1, "is_sales_item": 1,
+			"is_purchase_item": 1})
+			
+		po = create_purchase_order(item_code=item.item_code, qty=1)
+		
+		self.assertEquals(get_ordered_qty(item_code=item.item_code, warehouse="_Test Warehouse - _TC"), 1)
```
Comment (nabinhait) [inline erpnext/buying/doctype/purchase_order/test_purchase_order.py]: Check with existing

File: erpnext/controllers/status_updater.py [hunk 32]
```diff
@@ -168,11 +171,12 @@ def update_qty(self, change_modified=True):
 			else:
 				args['cond'] = ' and parent!="%s"' % self.name.replace('"', '\"')
 
-			args['set_modified'] = ''
 			if change_modified:
 				args['set_modified'] = ', modified = now(), modified_by = "{0}"'\
 					.format(frappe.db.escape(frappe.session.user))
-
+			
+			args["drop_ship_cond"] = ''
```
Comment (nabinhait) [inline erpnext/controllers/status_updater.py]: remove

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 33]
```diff
@@ -15,9 +15,21 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
...
-
+		var is_delivered_by_supplier = false;
+		var is_delivery_note = false;
+		
 		if(doc.docstatus==1) {
-			if(doc.status != 'Stopped') {
+			if(doc.status != 'Stopped' && doc.status != 'Closed') {
+				
+				$.each(cur_frm.doc.items, function(i, item){
+					if((item.delivered_by_supplier == 1 || item.supplier) && (item.qty > item.ordered_qty)){
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: Move `(item.qty > item.ordered_qty)` to inner loop

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 34]
```diff
@@ -25,7 +37,7 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
 				// 	doc.per_billed);
 
 				// indent
-				if(!doc.order_type || ["Sales", "Shopping Cart"].indexOf(doc.order_type)!==-1)
+				if(!doc.order_type || ["Sales", "Shopping Cart"].indexOf(doc.order_type)!==-1 && flt(doc.per_delivered, 2) < 100)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: && ! is_delivered_by_supplier

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 35]
```diff
@@ -33,27 +45,36 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
...
 
 				// stop
-				if(flt(doc.per_delivered, 2) < 100 || doc.per_billed < 100)
-					cur_frm.add_custom_button(__('Stop'), cur_frm.cscript['Stop Sales Order'])
-
-					// maintenance
-					if(flt(doc.per_delivered, 2) < 100 && ["Sales", "Shopping Cart"].indexOf(doc.order_type)===-1) {
-						cur_frm.add_custom_button(__('Maint. Visit'), this.make_maintenance_visit);
-						cur_frm.add_custom_button(__('Maint. Schedule'), this.make_maintenance_schedule);
+				if((flt(doc.per_delivered, 2) < 100 && is_delivery_note) || doc.per_billed < 100
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: flt(doc.per_delivered, 2) < 100 || flt(doc.per_billed) < 100

File: erpnext/selling/doctype/sales_order/sales_order.js [hunk 36]
```diff
@@ -33,27 +45,36 @@ erpnext.selling.SalesOrderController = erpnext.selling.SellingController.extend(
...
-					// sales invoice
-					if(flt(doc.per_billed, 2) < 100) {
-						cur_frm.add_custom_button(__('Invoice'), this.make_sales_invoice).addClass("btn-primary");
-					}
+				// sales invoice
+				if(flt(doc.per_billed, 2) < 100) {
+					cur_frm.add_custom_button(__('Invoice'), this.make_sales_invoice).addClass("btn-primary");
+				}
+				
+				if(flt(doc.per_ordered, 2) < 100 && is_delivered_by_supplier)
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.js]: per_ordered?

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 37]
```diff
@@ -252,6 +258,9 @@ def _valid_for_reserve(item_code, warehouse):
 
 	def on_update(self):
 		pass
+	
+	def before_update_after_submit(self):
+		self.validate_drop_ship()
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Validate if already PO exists

File: erpnext/selling/doctype/sales_order/sales_order.py [hunk 38]
```diff
@@ -488,3 +497,102 @@ def get_events(start, end, filters=None):
...
+				["delivery_date", "schedule_date"]
+			],
+			"field_no_map": [
+				"rate",
+				"price_list_rate"
+			],
+			"postprocess": update_item,
+			"condition": lambda doc: doc.ordered_qty < doc.qty and doc.supplier == for_supplier
+		},
+		"Sales Taxes and Charges": {
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/sales_order.py]: Don't map taxes

File: erpnext/selling/doctype/sales_order/test_sales_order.py [hunk 39]
```diff
@@ -295,7 +295,98 @@ def test_auto_insert_price(self):
...
+		
+		#test per_delivered status
+		delivered_by_supplier(po.name)
+		per_delivered = frappe.db.sql("""select sum(if(qty > ifnull(delivered_qty, 0), delivered_qty, qty))/sum(qty)*100 as per_delivered 
+			from `tabSales Order Item` where parent="{0}" """.format(so.name))
+
+		self.assertEquals(frappe.db.get_value("Sales Order", so.name, "per_delivered"), per_delivered[0][0])
+		
+		dn = create_dn_against_so(so.name, delivered_qty=1)
+
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/test_sales_order.py]: Test reserve qty again

File: erpnext/selling/doctype/sales_order/test_sales_order.py [hunk 40]
```diff
@@ -295,7 +295,98 @@ def test_auto_insert_price(self):
...
+		delivered_by_supplier(po.name)
+		per_delivered = frappe.db.sql("""select sum(if(qty > ifnull(delivered_qty, 0), delivered_qty, qty))/sum(qty)*100 as per_delivered 
+			from `tabSales Order Item` where parent="{0}" """.format(so.name))
+
+		self.assertEquals(frappe.db.get_value("Sales Order", so.name, "per_delivered"), per_delivered[0][0])
+		
+		dn = create_dn_against_so(so.name, delivered_qty=1)
+		
+		so.db_set('status', "Closed")
+		so.update_reserved_qty()
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/test_sales_order.py]: call set_status and check reserved and ordered qty again

File: erpnext/selling/doctype/sales_order/test_sales_order.py [hunk 41]
```diff
@@ -295,7 +295,98 @@ def test_auto_insert_price(self):
...
+		self.assertEquals(po.items[0].prevdoc_docname, so.name)
+		self.assertEquals(po.items[0].item_code, po_item.item_code)
+		self.assertEquals(dn.items[0].item_code, dn_item.item_code)
+		
+		#test ordered_qty and reserved_qty		
+		ordered_qty, reserved_qty = frappe.db.get_value("Bin", 
+			{"item_code": po_item.item_code, "warehouse": "_Test Warehouse - _TC"}, ["ordered_qty", "reserved_qty"])
+		
+		existing_ordered_qty = bin[0].ordered_qty if bin else 0.0
+		existing_reserved_qty = bin[0].reserved_qty if bin else 0.0
```
Comment (nabinhait) [inline erpnext/selling/doctype/sales_order/test_sales_order.py]: Check for both items

File: erpnext/buying/doctype/purchase_order/purchase_order.js (hunk 23 above)
Comment (rmehta) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]: Resume

File: erpnext/buying/doctype/purchase_order/purchase_order.js [hunk 42]
```diff
@@ -155,6 +161,29 @@ erpnext.buying.PurchaseOrderController = erpnext.buying.BuyingController.extend(
...
+	},
+	close_purchase_order: function(){
+		cur_frm.cscript.update_status('Close', 'Closed')
+	},
+	delivered_by_supplier: function(){
+		return frappe.call({
+			method: "erpnext.buying.doctype.purchase_order.purchase_order.delivered_by_supplier",
+			freez: true,
```
Comment (rmehta) [inline erpnext/buying/doctype/purchase_order/purchase_order.js]: freeze

**Pull request merged**